redis7/redis.go

```diff
@@ -0,0 +1,353 @@
+	return t.store
+}
+
+func (t tx) Unwrap() interface{} {
+	return t.writer
+}
+
+type shelf struct {
+	name   string
+	prefix string
+	reader redis.Cmdable
+	// writer holds the Redis command writer for writing to the database.
+	// It is separate from the reader because writes use a transactional Redis pipeline, which buffers the commands.
+	// Buffering read commands would result in unexpected behaviour where the reads are only executed on transaction commit.
+	writer redis.Cmdable
+	store  *store
+}
+
+func (s shelf) Put(key stoabs.Key, value []byte) error {
+	if err := s.store.checkOpen(); err != nil {
+		return err
+	}
+	return s.writer.Set(context.TODO(), s.toRedisKey(key), value, 0).Err()
+}
+
+func (s shelf) Delete(key stoabs.Key) error {
+	return s.writer.Del(context.TODO(), s.toRedisKey(key)).Err()
+}
+
+func (s shelf) Get(key stoabs.Key) ([]byte, error) {
+	result, err := s.reader.Get(context.TODO(), s.toRedisKey(key)).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return []byte(result), nil
+}
+
+func (s shelf) Iterate(callback stoabs.CallerFn) error {
+	var cursor uint64
+	var err error
+	var keys []string
+	for {
+		scanCmd := s.reader.Scan(context.TODO(), cursor, s.toRedisKey(stoabs.BytesKey("*")), resultCount)
+		keys, cursor, err = scanCmd.Result()
+		if err != nil {
+			return err
+		}
+		err := s.visitKeys(keys, callback)
+		if err != nil {
+			return err
+		}
+		if cursor == 0 {
+			// Done
+			break
+		}
+	}
+	return nil
+}
+
+func (s shelf) Range(from stoabs.Key, to stoabs.Key, callback stoabs.CallerFn) error {
+	keys := make([]string, 0, resultCount)
+	// Iterate from..to (start inclusive, end exclusive)
+	var numKeys = 0
+	for curr := from; bytes.Compare(curr.Bytes(), to.Bytes()) == -1; curr = curr.Next() {
+		keys = append(keys, s.toRedisKey(curr))
+		// We don't want to perform requests that are really large, so we limit it at resultCount keys
+		if numKeys >= resultCount {
+			err := s.visitKeys(keys, callback)
+			if err != nil {
+				return err
+			}
+			keys = make([]string, 0, resultCount)
+			numKeys = 0
+		} else {
+			numKeys++
+		}
+	}
+	return s.visitKeys(keys, callback)
+}
+
+func (s shelf) visitKeys(keys []string, callback stoabs.CallerFn) error {
+	values, err := s.reader.MGet(context.TODO(), keys...).Result()
+	if err != nil {
+		return err
+	}
+	for i, value := range values {
+		if values[i] == nil {
+			// Value does not exist (anymore), or not a string
+			continue
+		}
+		err := callback(s.fromRedisKey(keys[i]), []byte(value.(string)))
+		if err != nil {
+			// Callback returned an error, stop iterate and return it
+			return err
+		}
+	}
+	return nil
+}
+
+func (s shelf) Stats() stoabs.ShelfStats {
+	return stoabs.ShelfStats{
+		NumEntries: 0,
+		ShelfSize:  0,
+	}
+}
+
+func (s shelf) toRedisKey(key stoabs.Key) string {
+	// TODO: Does string(key) work for all keys? Especially when some kind of guaranteed ordering is expected?
```

Comment:
Redis doesn't have ordering. Ordering comes from `key.Next()`.
You have to call `key.String()` for the correct string representation.